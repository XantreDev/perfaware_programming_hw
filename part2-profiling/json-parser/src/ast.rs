use std::{fmt::Debug, iter::Peekable};

use crate::lexer::{Token, TokenStream};

#[derive(Debug)]
pub struct KeyValuePair(pub String, pub Ast);

#[derive(Debug)]
pub enum Ast {
    Object(Vec<KeyValuePair>),
    Array(Vec<Ast>),
    Number(f64),
    Bool(bool),
    Null,
    String(String),
}

pub(crate) struct InvalidTokenStreamError {
    pub(crate) message: String,
}

impl InvalidTokenStreamError {
    fn new(message: &str) -> InvalidTokenStreamError {
        InvalidTokenStreamError {
            message: message.to_string(),
        }
    }
    fn unexpected_end_of_tokens(location: &str) -> InvalidTokenStreamError {
        InvalidTokenStreamError {
            message: format!("unexpected end of tokens in {}", location),
        }
    }

    fn unexpected_token<T: Debug>(actual: T, expected: &str) -> InvalidTokenStreamError {
        InvalidTokenStreamError {
            message: format!("unexpected token: {:?} ({} is expected) ", actual, expected),
        }
    }
}

fn parse_object<T: Iterator<Item = Token>>(
    iter: &mut Peekable<T>,
) -> Result<Ast, InvalidTokenStreamError> {
    let mut content = Vec::new();

    loop {
        let Some(next) = iter.peek() else {
            return Err(InvalidTokenStreamError::unexpected_end_of_tokens("object"));
        };

        match next {
            Token::BraceClose => {
                return Ok(Ast::Object(content));
            }
            Token::String(key) => {
                let key = key.clone();
                iter.next();
                let colon = iter.next();
                if !matches!(colon, Some(Token::Colon)) {
                    return Err(InvalidTokenStreamError {
                        message: format!("unexpected token: {:?} (colon is expected) ", colon),
                    });
                }

                let ast_node = parse_unknown(iter)?;

                content.push(KeyValuePair(key, ast_node));
                let next_token = iter.peek();
                match next_token {
                    Some(Token::Comma) => {
                        iter.next();
                    }
                    Some(Token::BraceClose) => {
                        continue;
                    }
                    _ => {
                        return Err(InvalidTokenStreamError::unexpected_token(
                            colon,
                            "comma or BraceClose",
                        ));
                    }
                }
            }
            _ => {
                return Err(InvalidTokenStreamError::unexpected_token(next, "String"));
            }
        }
    }
}

fn parse_array<T: Iterator<Item = Token>>(
    iter: &mut Peekable<T>,
) -> Result<Ast, InvalidTokenStreamError> {
    let mut content = Vec::new();

    loop {
        let Some(next_token) = iter.peek() else {
            return Err(InvalidTokenStreamError::unexpected_end_of_tokens("array"));
        };

        match next_token {
            Token::BracketClose => {
                return Ok(Ast::Array(content));
            }
            _ => {
                let ast_node = parse_unknown(iter)?;

                content.push(ast_node);
                let Some(next_token) = iter.peek() else {
                    return Err(InvalidTokenStreamError::unexpected_end_of_tokens("array"));
                };

                match next_token {
                    Token::Comma => {
                        iter.next();
                    }
                    Token::BracketClose => {
                        continue;
                    }
                    _ => {
                        return Err(InvalidTokenStreamError::unexpected_token(
                            next_token,
                            "BracketClose or Comma",
                        ));
                    }
                }
            }
        }
    }
}

fn parse_number(number_str: &String) -> Result<f64, InvalidTokenStreamError> {
    let result = number_str.parse::<f64>();
    return result.map_err(|_| InvalidTokenStreamError::new("invalid fp number"));
}

fn parse_unknown<T: Iterator<Item = Token>>(
    iter: &mut Peekable<T>,
) -> Result<Ast, InvalidTokenStreamError> {
    let Some(next_token) = iter.next() else {
        return Err(InvalidTokenStreamError {
            message: "unexpected token stream end".to_string(),
        });
    };
    let ast_node = match next_token {
        Token::Bool(bool) => Ast::Bool(bool),
        Token::Comma => panic!("unexpected comma"),
        Token::Null => Ast::Null,
        Token::String(str) => Ast::String(str.to_owned()),
        Token::Number(value) => Ast::Number(parse_number(&value)?),
        Token::BraceOpen => {
            let obj = parse_object(iter)?;
            let next_token = iter.next();
            if !matches!(next_token, Some(Token::BraceClose)) {
                return Err(InvalidTokenStreamError {
                    message: format!("brace close expected, but got {:?}", next_token),
                });
            }

            obj
        }
        Token::BracketOpen => {
            let array = parse_array(iter)?;

            let next_token = iter.next();

            if !matches!(next_token, Some(Token::BracketClose)) {
                return Err(InvalidTokenStreamError {
                    message: format!("bracket close expected, but got {:?}", next_token),
                });
            }

            array
        }
        Token::BraceClose | Token::BracketClose | Token::Colon => {
            return Err(InvalidTokenStreamError {
                message: format!("invariant {:?} is unexpected", next_token),
            });
        }
    };

    Ok(ast_node)
}

pub fn build_ast(token_stream: TokenStream) -> Result<Ast, InvalidTokenStreamError> {
    let mut iter = token_stream.tokens.into_iter().peekable();

    parse_unknown(&mut iter)
}
