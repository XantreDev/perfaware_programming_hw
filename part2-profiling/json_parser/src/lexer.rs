use std::char;

use crate::ast::ParseError;

#[derive(Debug, Clone)]
pub enum Token {
    BraceOpen,
    BraceClose,
    BracketOpen,
    BracketClose,
    Null,
    Bool(bool),
    String(String),
    Number(f64),
    Comma,
    Colon,
}
// TODO: probably arena allocations will be beneficial
fn is_whitespace(value: char) -> bool {
    value == ' ' || value == '\n' || value == '\r' || value == '\t'
}

#[derive(Debug)]
pub struct TokenStream {
    pub tokens: Vec<Token>,
}

fn skip_n<T: Iterator>(value: &mut T, elements: u32) {
    for _ in 0..elements {
        value.next();
    }
}

pub(crate) struct Lexer<'a> {
    data: &'a String,
    position: usize,
}

impl Lexer<'_> {
    pub(crate) fn new<'a>(value: &'a String) -> Lexer<'a> {
        Lexer {
            data: value,
            position: 0,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Result<Token, ParseError>;

    fn next(&mut self) -> Option<Self::Item> {
        lexicize(self)
    }
}

#[inline(always)]
fn get_current_char<'a>(lexer: &mut Lexer<'a>) -> Option<char> {
    lexer
        .data
        .get(lexer.position..=lexer.position)
        .and_then(|it| it.chars().next())
}

#[inline(always)]
fn skip_char<'a>(lexer: &mut Lexer<'a>) {
    lexer.position += 1;
}

#[inline(always)]
fn get_current_char_and_skip<'a>(lexer: &mut Lexer<'a>) -> Option<char> {
    let char = get_current_char(lexer);

    lexer.position += 1;
    char
}

trait AsciiReadable {
    fn read_ascii_char_at(&self, idx: usize) -> Option<char>;
}
impl AsciiReadable for str {
    fn read_ascii_char_at(&self, idx: usize) -> Option<char> {
        self.get(idx..=idx).unwrap().chars().next()
    }
}
impl AsciiReadable for String {
    fn read_ascii_char_at(&self, idx: usize) -> Option<char> {
        self.get(idx..=idx).unwrap().chars().next()
    }
}

pub fn lexicize(lexer: &mut Lexer<'_>) -> Option<Result<Token, ParseError>> {
    loop {
        if lexer.data.len() <= lexer.position {
            return None;
        };
        let Some(char) = get_current_char(lexer) else {
            return None;
        };
        if is_whitespace(char) {
            lexer.position += 1;
        } else {
            break;
        }
    }

    let Some(char) = get_current_char_and_skip(lexer) else {
        return None;
    };

    let token = match char {
        '{' => Token::BraceOpen,
        ':' => Token::Colon,
        '}' => Token::BraceClose,
        '[' => Token::BracketOpen,
        ']' => Token::BracketClose,
        'n' => {
            match consume_str(lexer, "ull") {
                Err(err) => return Some(Err(err)),
                _ => {}
            };
            Token::Null
        }
        ',' => Token::Comma,
        't' => {
            match consume_str(lexer, "rue") {
                Err(err) => return Some(Err(err)),
                _ => {}
            };
            Token::Bool(true)
        }
        'f' => {
            match consume_str(lexer, "alse") {
                Err(err) => return Some(Err(err)),
                _ => {}
            };

            Token::Bool(false)
        }
        '"' => {
            let string_token = parse_string(lexer);
            match string_token {
                Err(res) => return Err(res).into(),
                Ok(res) => Token::String(res),
            }
        }
        _ if is_zero_nine_digit(&char) || char == '-' => {
            let res = parse_number(&char, lexer);
            match res {
                Err(res) => return Err(res).into(),
                Ok(res) => Token::Number(res),
            }
        }
        _ => {
            return Err(ParseError::from_string(format!("invariant char {}", char))).into();
        }
    };

    Some(Ok(token))
}

fn lexicize_complete(data: String) -> Result<TokenStream, ParseError> {
    let mut lexer = Lexer {
        data: &data,
        position: 0,
    };

    let mut tokens = Vec::new();

    loop {
        let Some(result) = lexicize(&mut lexer) else {
            return Ok(TokenStream { tokens: tokens });
        };
        tokens.push(result?);
    }
}

fn consume_str<'a>(lexer: &mut Lexer<'a>, to_consume: &str) -> Result<(), ParseError> {
    let mut idx = 0;

    while idx < to_consume.len() {
        let Some(char) = get_current_char_and_skip(lexer) else {
            return Err(ParseError::new("unexpected tokens end"));
        };
        let expected_char = to_consume.read_ascii_char_at(idx).unwrap();
        if expected_char != char {
            return Err(ParseError::from_string(format!(
                "exected {}, but got {}",
                expected_char, char
            )));
        }

        idx += 1;
    }

    return Ok(());
}

fn is_one_nine_digit(char: &char) -> bool {
    matches!(char, '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9')
}
fn is_zero_nine_digit(char: &char) -> bool {
    matches!(
        char,
        '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9'
    )
}

fn parse_number(starting_char: &char, lexer: &mut Lexer<'_>) -> Result<f64, ParseError> {
    let sign: f64 = if starting_char.to_owned() == '-' {
        -1.0
    } else {
        1.0
    };

    let mut integer: u64 = 0;
    if is_one_nine_digit(starting_char) {
        integer = (*starting_char as u64) - ('0' as u64);
    }

    loop {
        let Some(char) = get_current_char(lexer) else {
            break;
        };
        if is_zero_nine_digit(&char) {
            integer = integer * 10 + ((char as u64) - ('0' as u64));
            skip_char(lexer);
        } else {
            break;
        }
    }

    let mut fraction: f64 = 0.0;
    if matches!(get_current_char(lexer), Some('.')) {
        skip_char(lexer);

        let mut mult = 0.1;
        let mut is_first_iter = true;

        loop {
            let Some(char) = get_current_char(lexer) else {
                break;
            };

            if is_zero_nine_digit(&char) {
                is_first_iter = false;
                skip_char(lexer);
                fraction = ((char as u64) - ('0' as u64)) as f64 * mult;
                mult = mult / 10.0;
            } else {
                break;
            }
        }

        if is_first_iter {
            return Err(ParseError::from_string(format!(
                "expected to have a least one digit in fraction part, but got {:?}",
                get_current_char(lexer)
            )));
        };
    }
    let mut exponent: i64 = 0;
    if matches!(get_current_char(lexer), Some('e') | Some('E')) {
        skip_char(lexer);
        let Some(char) = get_current_char(lexer) else {
            return Err(ParseError::new("unexpected end of input in exponent form"));
        };
        let sign = if char == '-' {
            skip_char(lexer);
            -1
        } else if char == '+' {
            skip_char(lexer);
            1
        } else {
            1
        };

        let mut is_first_iter = true;
        loop {
            let Some(char) = get_current_char(lexer) else {
                break;
            };

            if (is_first_iter && !is_one_nine_digit(&char)) || !is_zero_nine_digit(&char) {
                break;
            }

            is_first_iter = false;
            skip_char(lexer);
            exponent = (exponent * 10) + ((char as i64) - ('0' as i64));
        }
        exponent = exponent * sign;

        if is_first_iter {
            return Err(ParseError::new(
                "expected to have a least one digit in fraction part",
            ));
        };
    }

    Ok((sign as f64) * ((integer as f64) + fraction) * ((10f64).powf(exponent as f64)))
}

fn parse_string(lexer: &mut Lexer<'_>) -> Result<String, ParseError> {
    let mut str = String::new();
    loop {
        let Some(char) = get_current_char_and_skip(lexer) else {
            return Err(ParseError::new("unexpected end of json"));
        };

        if char == '"' {
            return Ok(str);
        }

        // escaping
        if char == '\\' {
            let Some(next_char) = get_current_char_and_skip(lexer) else {
                return Err(ParseError::new("unexpected end by escape character"));
            };

            str.push(next_char);
            continue;
        }

        str.push(char);
    }
}

#[test]
fn check_basic_lexing() {
    insta::assert_debug_snapshot!(lexicize_complete("{ \"about\": 10 }".to_string()).unwrap());
}
#[test]
fn check_basic_lexing2() {
    insta::assert_debug_snapshot!(lexicize_complete("123.4".to_string()).unwrap());
}

#[test]
fn check_basic_lexing_exponential() {
    insta::assert_debug_snapshot!(
        lexicize_complete(
            "[123.4e1, 123e1, 123.04e1, 123.04E1, 123.04E-1, 123.04E-12, 0.5e-12]]".to_string()
        )
        .unwrap()
    );
}

#[test]
fn check_basic_lexing3() {
    insta::assert_debug_snapshot!(
        lexicize_complete(
            "{
            \"ability\": [1, null, false, 2.0, \"55\", { \"obj\": 213 }],
            \"key\": { \"value\": 220 }
        }"
            .to_string()
        )
        .unwrap()
    );
}
