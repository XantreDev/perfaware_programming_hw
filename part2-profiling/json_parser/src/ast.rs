use std::{fmt::Debug, iter::Peekable};

use crate::lexer::{Token, TokenStream};

#[derive(Debug)]
pub struct KeyValuePair(pub String, pub Ast);

#[derive(Debug)]
pub enum Ast {
    Object(Vec<KeyValuePair>),
    Array(Vec<Ast>),
    Number(f64),
    Bool(bool),
    Null,
    String(String),
}

#[derive(Debug, Clone)]
pub struct ParseError {
    pub message: String,
}

impl ParseError {
    pub(crate) fn from_string(message: String) -> ParseError {
        ParseError { message: message }
    }

    pub(crate) fn new(message: &str) -> Self {
        ParseError {
            message: message.to_string(),
        }
    }

    pub(crate) fn unexpected_end_of_tokens(location: &str) -> ParseError {
        ParseError {
            message: format!("unexpected end of tokens in {}", location),
        }
    }

    pub(crate) fn unexpected_token<T: Debug>(actual: T, expected: &str) -> ParseError {
        ParseError {
            message: format!("unexpected token: {:?} ({} is expected) ", actual, expected),
        }
    }
}

fn parse_object<T: Iterator<Item = Result<Token, ParseError>>>(
    iter: &mut Peekable<T>,
) -> Result<Ast, ParseError> {
    let mut content = Vec::new();

    loop {
        let Some(next) = iter.peek() else {
            return Err(ParseError::unexpected_end_of_tokens("object"));
        };
        let next = match next {
            Ok(it) => it,
            Err(err) => return Err(err.clone()),
        };

        match next {
            Token::BraceClose => {
                return Ok(Ast::Object(content));
            }
            Token::String(key) => {
                let key = key.clone();
                iter.next();
                match iter.next() {
                    Some(Ok(Token::Colon)) => {}
                    Some(Ok(token)) => {
                        return Err(ParseError {
                            message: format!("unexpected token: {:?} (colon is expected) ", token),
                        });
                    }
                    Some(Err(err)) => return Err(err.clone()),
                    None => {
                        return Err(ParseError::unexpected_end_of_tokens("object"));
                    }
                };

                let ast_node = parse_unknown(iter)?;

                content.push(KeyValuePair(key, ast_node));
                match iter.peek() {
                    Some(Ok(Token::Comma)) => {
                        iter.next();
                    }
                    Some(Ok(Token::BraceClose)) => {
                        continue;
                    }
                    Some(Ok(token)) => {
                        return Err(ParseError::unexpected_token(token, "comma or BraceClose"));
                    }
                    Some(Err(_)) => return Err(iter.next().unwrap().unwrap_err()),
                    None => return Err(ParseError::unexpected_end_of_tokens("object")),
                }
            }
            _ => {
                return Err(ParseError::unexpected_token(next, "String"));
            }
        }
    }
}

fn parse_array<T: Iterator<Item = Result<Token, ParseError>>>(
    iter: &mut Peekable<T>,
) -> Result<Ast, ParseError> {
    let mut content = Vec::new();

    loop {
        let Some(next_token) = iter.peek() else {
            return Err(ParseError::unexpected_end_of_tokens("array"));
        };

        match next_token {
            Err(err) => return Err(err.clone()),
            Ok(Token::BracketClose) => {
                return Ok(Ast::Array(content));
            }
            _ => {
                let ast_node = parse_unknown(iter)?;

                content.push(ast_node);
                let Some(next_token) = iter.peek() else {
                    return Err(ParseError::unexpected_end_of_tokens("array"));
                };

                match next_token {
                    Err(err) => return Err(err.clone()),
                    Ok(Token::Comma) => {
                        iter.next();
                    }
                    Ok(Token::BracketClose) => {
                        continue;
                    }
                    Ok(next_token) => {
                        return Err(ParseError::unexpected_token(
                            next_token,
                            "BracketClose or Comma",
                        ));
                    }
                }
            }
        }
    }
}

pub(crate) fn parse_unknown<T: Iterator<Item = Result<Token, ParseError>>>(
    iter: &mut Peekable<T>,
) -> Result<Ast, ParseError> {
    let Some(next_token) = iter.next() else {
        return Err(ParseError::new("unexpected token stream end"));
    };
    let next_token = next_token?;
    let ast_node = match next_token {
        Token::Bool(bool) => Ast::Bool(bool),
        Token::Comma => return Err(ParseError::new("unexpected comma")),
        Token::Null => Ast::Null,
        Token::String(str) => Ast::String(str.to_owned()),
        Token::Number(value) => Ast::Number(value),
        Token::BraceOpen => {
            let obj = parse_object(iter)?;
            let next_token = iter.next();
            if !matches!(next_token, Some(Ok(Token::BraceClose))) {
                return Err(ParseError::from_string(format!(
                    "brace close expected, but got {:?}",
                    next_token
                )));
            }

            obj
        }
        Token::BracketOpen => {
            let array = parse_array(iter)?;

            let next_token = iter.next();

            if !matches!(next_token, Some(Ok(Token::BracketClose))) {
                return Err(ParseError::from_string(format!(
                    "bracket close expected, but got {:?}",
                    next_token
                )));
            }

            array
        }
        Token::BraceClose | Token::BracketClose | Token::Colon => {
            return Err(ParseError::from_string(format!(
                "invariant {:?} is unexpected",
                next_token
            )));
        }
    };

    Ok(ast_node)
}
