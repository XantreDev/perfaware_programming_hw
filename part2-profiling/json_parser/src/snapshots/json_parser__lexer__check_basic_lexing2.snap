---
source: src/lexer.rs
expression: "lexicize_complete(\"123.4\".to_string()).unwrap()"
---
TokenStream {
    tokens: [
        Number(
            123.4,
        ),
    ],
}
